src/Simplex/Messaging/Agent.hs

```diff
@@ -1400,13 +1403,20 @@ enqueueMessageB c reqs = do
           agentMsgStr = smpEncode agentMsg
           internalHash = C.sha256Hash agentMsgStr
           currentE2EVersion = maxVersion e2eEncryptVRange
+      -- TODO [save once] Save single MsgBody / enveloped body agentMsgStr (outside of withStoreBatch ... storeSentMsg).
+      -- TODO             Link all messages (message deliveries) to it, save encryption data per message, e.g. pqEnc_ (intent, not result).
+      -- TODO             'msg_body' field is not nullable - use default empty strings?
+      -- TODO             No need to save specific ratchet state/keys for encryption at this point? - Get them at point of delivery?
```

Comment:
There are probably two options:
1. save ratchet keys at the point of scheduling delivery - this is probably simpler
2. save ratchet keys on the first delivery attempt - the advantage here is that undelivered messages won't be "wasting" IDs and the recipient has less to catch up with on expired messages. The disadvantage here is that there may need to be some other way to signal skipped messages. Although if we use internal IDs for skipped and ratchet numbering is independent it's probably ok.

In any case, there is no option to avoid saving keys / ratchet IDs / headers etc., because they can't change between attempts.